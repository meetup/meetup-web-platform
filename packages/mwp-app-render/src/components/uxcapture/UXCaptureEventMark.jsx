// @flow
import * as React from 'react';

type onLoadHTMLElements =
	| React.Element<'iframe'>
	| React.Element<'img'>
	| React.Element<'input'>;

type Props = {
	mark: string,
	children: onLoadHTMLElements,
};

/**
 * takes only a *single* child element
 * and injects event based UX.mark() call into the child
 *
 * @see example https://github.com/meetup/ux-capture#image-elements
 */
const UXCaptureEventMark = ({ mark, children }: Props) => {
	// if child has on onLoad prop
	// don't do anything
	// TODO: we may want to merge onLoad props instead of ignoring
	if (children.props.onLoad) {
		return children;
	}

	return React.cloneElement(children, {
		onLoad: `if(window.UX) { UX.mark('${mark}'); }`,
	});
};

export default UXCaptureEventMark;
