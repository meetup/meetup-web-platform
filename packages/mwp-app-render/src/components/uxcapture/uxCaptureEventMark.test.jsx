import React from 'react';
import { shallow } from 'enzyme';
import UXCaptureEventMark from './UXCaptureEventMark';

const MOCK_PROPS = {
	mark: 'ux-image-load-logo',
};

const MOCK_CHILD = <img src="/foo" />;

const renderComponent = (props = MOCK_PROPS, children = MOCK_CHILD) =>
	shallow(
		<UXCaptureEventMark {...props}>
			{children}
		</UXCaptureEventMark>
	);

describe('UXCaptureEventMark', () => {
	it('renders component markup', () => {
		expect(renderComponent()).toMatchSnapshot();
	});

	it('returns the child element untouched if it has an onLoad prop', () => {
		const onLoadChild = <img src="/foo" onLoad="alert('hi');" />;
		expect(renderComponent(MOCK_PROPS, onLoadChild)).toMatchSnapshot();
	});
});
