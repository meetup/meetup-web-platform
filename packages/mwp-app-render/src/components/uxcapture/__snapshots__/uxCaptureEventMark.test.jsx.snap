// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UXCaptureEventMark renders component markup 1`] = `
<img
  onLoad="if(window.UX) { UX.mark('ux-image-load-logo'); }"
  src="/foo"
/>
`;

exports[`UXCaptureEventMark returns the child element untouched if it has an onLoad prop 1`] = `
<img
  onLoad="alert('hi');"
  src="/foo"
/>
`;
